### PR TITLE
preserve x_ phases and types in all CPAN::Meta::Prereqs interfaces

### PR DESCRIPTION
**Pull request by karenetheridge**
as_string_hash and with_merged_prereqs were not respecting these extra fields.

### CODE REVIEW

**Comment by karenetheridge** [comment]
fixing travis failure now... hang tight...

**Comment by kentfredric** [comment]
Probably going to want a hook in dzil eventually to make sure this works when used. Yay.

**Comment by karenetheridge** [comment]
@kentfredric I was going to simply submit a PR to bump the dep in Dist::Zilla::Prereqs to the latest CPAN::Meta::Prereqs, rather than switching on versions. :)

**Comment by kentfredric** [comment]
This issue closes #66 

Yay. 2014...

File: lib/CPAN/Meta/Prereqs.pm [hunk 1]
```diff
@@ -15,6 +15,7 @@ organized by phase and type, as described in L<CPAN::Meta::Prereqs>.
 
 use Carp qw(confess);
 use Scalar::Util qw(blessed);
+use List::Util 1.45 qw(uniq);
```
Comment (haarg) [inline lib/CPAN/Meta/Prereqs.pm]: I don't think requiring a recent XS module is a good idea given this module's position in the toolchain.

**Comment by karenetheridge** [comment]
new commit added to make these interfaces public. I'm not entirely sure that I picked the best names though -- is `my @phases = $prereqs->phases;` and `my @runtime_types = $prereqs->types_in('runtime');` clear enough?  `types_for` might be better than `types_in`?

**Comment by karenetheridge** [comment]
@haarg I'll remove the usage of List::Util::uniq.  Is this a suitable replacement?

```
sub __uniq {
  goto &List::Util::uniq if \&List::Util::uniq;
  keys %{ +{ map { $_ => undef } @_ } }  # from List::MoreUtils::PP
}
```

**Comment by haarg** [comment]
I'd go with
```
sub _uniq {
  my (%s, $u);
  grep { defined ? !$s{$_}++ : !$u++ } @_;
}
```
just to keep consistency, even if we don't need to handle undef or maintain ordering.

**Comment by xdg** [comment]
Squashed and merged.  Thanks!

**Pull request merged**